### PR TITLE
Enable type-safe interoperability between different independent Python/C++ bindings systems.

### PR DESCRIPTION
**Pull request by rwgk**
## Description
This PR enables type-safe interoperability between different independent Python/C++ bindings systems, including pybind11 versions with different `PYBIND11_INTERNALS_VERSION`s. That means, innovations that require pybind11 internals version bumps become far less disruptive, almost to a negligible degree for most practical purposes (see **NOTE** below for the limitation).

More concretely, this PR enables extracting raw C++ pointers from Python instances of `py::class_`-wrapped types.

The implementation involves a
```python
def _pybind11_conduit_v1_(
    self,
    pybind11_platform_abi_id: bytes,
    cpp_type_info_capsule: capsule,
    pointer_kind: bytes) -> capsule
```
method that is added *unconditionally* to all `py::class_`-wrapped types (to avoid "oh, too late!" situations). — The implementation of this method is identical for all types, i.e. it is not templated, therefore it compiles very quickly and adds very little binary size overhead. I.e. the cost is very small, while the benefit of maximizing interoperability is expected to be very large.

The design of the `_pybind11_conduit_v1_` feature provides two layers of protection against C++ ABI mismatches:

* The first and most important layer is that the `pybind11_platform_abi_id`s must match between extensions. — This will never be perfect, but is the same pragmatic approach [used in pybind11 since 2017](https://github.com/pybind/pybind11/commit/96997a4b9d4ec3d389a570604394af5d5eee2557) (`PYBIND11_INTERNALS_ID`). 

* The second layer is that the `typeid(std::type_info).name()`s must match between extensions.

This PR includes calls of the `_pybind11_conduit_v1_()` method from:

* `pybind11::detail::type_caster_generic` (in pybind11/detail/type_caster_base.h, with helper functions in pybind11/detail/cpp_conduit.h)

* tests/exo_planet_c_api.cpp

The latter is a great way to quickly learn how the `_pybind11_conduit_v1_` feature works. exo_planet_c_api.cpp only uses the plain CPython API, not any pybind11 features, except one macro resolving to a C++ string literal: `PYBIND11_PLATFORM_ABI_ID`.

The integration into `type_caster_generic` means that any object, from any Python/C++ bindings system (e.g. **in theory at the moment** SWIG, Cython, PyCLIF, nanobind) providing a compatible `self._pybind11_conduit_v1_()` method can be passed from Python to pybind11-wrapped C++ functions.

For entertainment: This PR makes pirate-like approaches to getting to the wrapped C++ pointers obsolete. [Here is a particularly creative piece of code](https://github.com/rwgk/stuff/blob/20912090ce615e96cf83088b7d55b6f931704b54/museum_of_codiosities/GetPyWrappedObjectPtr.hpp) enabling SWIG-pybind11 interoperability, exactly as found in the wild, in an undisclosed location.

**NOTE: Matching `PYBIND11_INTERNALS_ID`s are still critically important for this situation:**

* `py::class_<base>` in one extension,
* `py::class_<derived, base>` in another.

For a demonstration, please see `test_home_planet_wrap_very_lonely_traveler()` and `test_exo_planet_pybind11_wrap_very_lonely_traveler()` in tests/test_cpp_conduit.py.
________

This PR **passed Google-global testing**:

* merged into the smart_holder branch: internal ID OCL:669841401:BASE:669907294:1725198946622:db3091a3
* [merged into pybind11clif](https://github.com/google/pybind11clif/pull/30158), then running with PyCLIF-pybind11: internal ID OCL:669954247:BASE:670062753:1725250171762:cb476a31
________

Suggestions for reviewing this PR (or just understanding what it does in detail):

* Start with test/exo_planet_c_api.cpp

* Review `cpp_conduit_method()` in pybind11/detail/type_caster_base.h, which is the only implementation in this PR building an object to be returned by `self._pybind11_conduit_v1_()`. — Other bindings systems would have their own implementations.

* The most interesting part of this PR is this small code fragment in `cpp_conduit_method()`:

```C++
    const auto *cpp_type_info = cap_cpp_type_info.get_pointer<const std::type_info>();
    type_caster_generic caster(*cpp_type_info);
    if (!caster.load(self, false)) {
        return none();
    }
```

**Note that this takes care of all inheritance-related pointer conversions (up- and down-casts).**

* Review the rest of the changes in include/pybind11/detail/type_caster_base.h, which calls helper functions in:

* include/pybind11/detail/cpp_conduit.h

* Review the very simple change in include/pybind11/detail/internals.h, which just factors out `PYBIND11_PLATFORM_ABI_ID` from `PYBIND11_INTERNALS_ID`.

* Review the very simple changes in include/pybind11/pybind11.h. — Note that the auto-magically added `_pybind11_conduit_v1_` method is not included in docstring signatures, to not clutter the docstrings.

* Review the added tests.
________

**Beyond this PR:**
________

Please take a quick look at the top of tests/exo_planet_c_api.cpp.

It is unfortunate that bindings systems other than pybind11 need to `#include <pybind11/pybind11.h>` just to get to the `PYBIND11_PLATFORM_ABI_ID` macro (which resolves to a simple C++ string literal). This leads to significant problems in environments that have C++ exception handling disabled (e.g. via `-fno-exceptions`).

An attempt was made to establish an alternative that maximizes reusability, but the implementation requires refactoring of pybind11/detail/common.h and pybind11/detail/internals.h. One particular difficulty is that `PYBIND11_PLATFORM_ABI_ID` can only be built after `#include <Python.h>`, because that transitively `#include`s system headers that define some of the system macros used in building `PYBIND11_PLATFORM_ABI_ID`. To not make this fairly simple PR significantly more complicated, the changes for the attempt were backed out. The pure refactoring work is best handled in a follow-on PR.

FWIW: In an ideal world, `PYBIND11_PLATFORM_ABI_ID` would be spelled as `PY_CPP_COMMUNITY_PLATFORM_ABI_ID` or similar, and come in a tiny additional header file included with CPython releases, so that all Python/C++ bindings systems could standardize on it.
________

Currently `b"raw_pointer_ephemeral"` (see above) is the only supported `pointer_kind`. Ideas for the future go in the direction of `b"std::shared_ptr"`, `b"std::unique_ptr"`, which will probably be relatively simple to add after the [smart_holder branch is merged](https://github.com/pybind/pybind11/pull/5339). — It is not clear (to rwgk@) if those additions will be worth the effort and added complexity, but having `pointer_kind` from the start is expressive and leaves the door open for future developments. — Additionally, the `pointer_kind` could be leveraged for implementing a `b"__diagnose__"` (or similar) feature, to report why retrieving a desired C++ pointer is failing.
________

**mypy stubgenc** will generate this additional signature:

```python
    def _pybind11_conduit_v1_(self, *args, **kwargs) -> Any: ...
```

If this is undesired by the community (?), the signature could be suppressed similar to how dunder methods are suppressed [here](https://github.com/python/mypy/blob/0a6d40b125f6b353621016d8a9022212104f7877/mypy/stubutil.py#L552).
________

For completeness, the starting point for this PR was PR #3633 & PR #3788, although this PR is very different in many details.

## Suggested changelog entry:

```rst
A new ``self._pybind11_conduit_v1_()`` method is automatically added to all ``py::class_``-wrapped types, to enable type-safe interoperability between different independent Python/C++ bindings systems, including pybind11 versions with different ``PYBIND11_INTERNALS_VERSION``s.
```

### CODE REVIEW

**Comment by wjakob** [comment]
Dear @rwgk,

thank you for looking into this. I think that an ABI for interoperability is indeed needed, especially for big binding efforts that are spread out over many extensions compiled with different frameworks. A precedent that has IMO been quite successful is DLPack (https://github.com/dmlc/dlpack), which was able to get various tensor libraries to talk to each other. 

One high-level comment is that for this kind of proposal to become successful and universally adapted, it will need more discussion from all the different stakeholders (think of a PEP in Python-land). Besides this point, this current proposal is scoped narrowly to have interop with pybind11 flavors. It has pybind11-named parameters and relies on internals of the library. To be a candidate for adoption by other binding tools, it will need a full and self-contained specification that isn't tied to pybind11.

Just from reading this writeup, there are a few points I'm thinking would be better to work out in detail:

1. Object ownership will need to be discussed in more detail. pybind11k is centered around holders (`unique_ptr`, `shared_ptr`), but other frameworks use different methods to deal with ownership. What does a to this conversion layer imply for who is allowed to access the object, and for how long?

2. If two separate Python extensions can exchange instances, it likely means that the type has been bound twice, which means that there are two incompatible type objects. What are the implications of that?

3. Can the ABI tag be defined in a self-contained way, in such a way as not to depend on pybind11?

4. What about locking and free-threading? Does it have an impact on this kind of ABI? An ABI is always a long-term commitment, so it's important to get it right the first time.

5. I'm not a big fan of specifying `pointer_type` as a character string (`bytes`) since there are overheads to creating and decoding them. Why not have a `mode` integer parameter with an enumeration explaining what is being exchanged? The proposal could initially define mode 0, with other ones being available later on.

6. AFAIK the proposal never specifies the full signature of ``__cpp_conduit__`` -- i.e., what it is expected to return (what Python type).

7. Can the operation ever fail? When should it fail, and how should that be handled by the caller?

8. It may be wise to add a `None`-initialized `payload` or `extra` parameter that could be used for future extensions of the API.

One thing to consider is whether it actually makes sense for ``_cpp_conduit__`` to be a Python function. It seems to me, that it is an implementation detail of the binding layer that no Python extension is ever expected to call itself. The DLPack specification takes a different approach here: it specifies a C data structure and C function pointers that are embedded into Python using a capsule. This benefit here is that a C function call is dramatically more efficient than the same call in Python, and this pays off when the "conduit" layer is on the critical path of two extensions that exchange calls at a high rate. It also keeps the implementation details out of Python.

These are just a few things I can think of. However, authors of other extension libraries (SWIG, Cython, etc.) may raise other points. I think that it might be wise to think longer-term and start a broader discussion rather than merging this proposal quickly as-is.

**Comment by wjakob** [comment]
Sorry, but this (aka. "the code is the specification" & "my way or the highway") is not the way that specifications connecting multiple complex frameworks are made, especially when this involves a long-term ABI contract for performance-critical functionality. 

I'm happy to discuss further, and I can also make another stab at a full specification. But I am strongly against rushing this.

**Comment by petersteneteg** [comment]
Would this also enable interaction with shiboken used by qt? 
https://doc.qt.io/qtforpython-6/shiboken6/index.html
Shiboken also have some similar functions 
Shiboken.getCppPointer, and Shiboken.wrapInstance. 
https://doc.qt.io/qtforpython-6/shiboken6/shibokenmodule.html

Might be interesting to compare a bit.

**Comment by rwgk** [comment]
> Would this also enable interaction with shiboken used by qt? https://doc.qt.io/qtforpython-6/shiboken6/index.html Shiboken also have some similar functions Shiboken.getCppPointer, and Shiboken.wrapInstance. https://doc.qt.io/qtforpython-6/shiboken6/shibokenmodule.html
> 
> Might be interesting to compare a bit.

Thanks @petersteneteg, that's something I wasn't aware off before.

From the documentation you linked:
________
* Shiboken.getCppPointer(obj)
    * Returns a **tuple of longs** that contain the memory addresses of the C++ instances wrapped by the given object.

* Shiboken.wrapInstance(address, type)
    * Creates a Python wrapper for a C++ object instantiated at a given memory address - the returned object type will be the same given by the user.

    * The type must be a Shiboken type, **the C++ object will not be destroyed** when the returned Python object reach zero references.

    * If the address is invalid or doesn’t point to a C++ object of given type the behavior is undefined.
________

My analysis, please correct me if I'm wrong:

Overall: There are no safety mechanisms at all (C++ type safety, C++ ABI safety).

**For `GetPointer()`:**

The implementation is here:

* https://github.com/pyside/Shiboken/blob/09e1ffb01597b6fe4ed1ba1ae362449e3780d7ed/shibokenmodule/typesystem_shiboken.xml#L29-L40

* https://github.com/pyside/Shiboken/blob/09e1ffb01597b6fe4ed1ba1ae362449e3780d7ed/libshiboken/basewrapper.cpp#L1008-L1015

That returns a tuple of `void *` cast to Python `int`. Only Shiboken knows which `int` corresponds to which C++ type. I.e. this does not have the information required for interoperability with other bindings systems. For example, the `load()` in a pybind11 `type_caster` could not use the tuple of `int`s.

**For `wrapInstance()`:**

The implementation is here:

* https://github.com/pyside/Shiboken/blob/09e1ffb01597b6fe4ed1ba1ae362449e3780d7ed/shibokenmodule/typesystem_shiboken.xml#L20-L27
* https://github.com/pyside/Shiboken/blob/09e1ffb01597b6fe4ed1ba1ae362449e3780d7ed/libshiboken/basewrapper.cpp#L1086-L1110

That's is a different beast altogether, because it is not for interoperability. It's basically a factory function that accept an arbitrary `unsigned long` and uses a C-style cast to `void *`, without any safety checks at all. It then builds a Python object **without taking ownership**, and **without establishing any formal lifetime relationship**. If the passed C++ pointer is deleted while the Python object is still alive, that's another undefined behavior situation.

We could implement something equivalent in pybind11, but I'd really rather not. It would essentially just be a fundamentally unsafe alternative to regular (and type-safe) pybind11 factory functions.

I'm wondering: What is the primary use of the `GetPointer()`, `wrapInstance()` functions? — Since C++ pointers are cast to Python `int`s and back, surely the pair of function calls must both run in the same process. Why would one want to go through a roundtrip, in the same process?

________

**Tangential**

Interestingly, there is also this:

* Shiboken.isValid(obj)

    * Given a Python object, returns True if the object methods can be called without an exception being thrown. A Python wrapper becomes invalid when the underlying C++ object is destroyed or unreachable.

It looks like Shiboken supports disowning. (Before I saw this I assumed only PyCLIF and pybind11 smart_holder can do that.)

File: include/pybind11/pybind11.h [hunk 1]
```diff
@@ -610,7 +610,8 @@ class cpp_function : public function {
         int index = 0;
         /* Create a nice pydoc rec including all signatures and
            docstrings of the functions in the overload chain */
-        if (chain && options::show_function_signatures()) {
+        if (chain && options::show_function_signatures()
+            && strcmp(rec->name, "__cpp_conduit__") != 0) {
```
Comment (Skylion007) [inline include/pybind11/pybind11.h:614]:
```suggestion
            && std::strcmp(rec->name, "__cpp_conduit__") != 0) {
```

File: include/pybind11/pybind11.h [hunk 2]
```diff
@@ -619,7 +620,7 @@ class cpp_function : public function {
         // Then specific overload signatures
         bool first_user_def = true;
         for (auto *it = chain_start; it != nullptr; it = it->next) {
-            if (options::show_function_signatures()) {
+            if (options::show_function_signatures() && strcmp(rec->name, "__cpp_conduit__") != 0) {
```
Comment (Skylion007) [inline include/pybind11/pybind11.h:623]:
```suggestion
            if (options::show_function_signatures() && std::strcmp(rec->name, "__cpp_conduit__") != 0) {
```

**Comment by ax3l** [comment]
This looks like a nice feature to add, e.g., for faster innovation where currently breaking changes are needed. The implementation looks low overhead and sensible to me on first few.

For the concern of long term compatibility:
If we added a spec version to the matched criteria for compatibility, and then tried it out in a few implementations (say nanobind, pybind11), we could experiment it while proposing it for a PEP.
This is the way array interface and cuda array interface were introduced and allows us to propose something we tried out in practice. We can also clearly mark it experimental for the first version.

**Comment by henryiii** [comment]
Would it be possible to make this internal (and maybe even experimental) for now while we work on a spec? Specifically, if this gets merged, then `__cpp_conduit__` becomes specified, and it's hard to change that specification. Also, only Python is officially allowed to use dunder methods; I know some libraries (like NumPy) also add some, but Python's stance is that all dunder names are reserved for Python's use. I think it would be okay to write something a little less official for one, but I don't think we should introduce one quite yet, especially without any specification .

I think if we made this `_pybind11_conduit_v1_`, for example, then we could have time to work on a proper written spec with buy-in from nanobind and hopefully others, with a more general name, while still getting the benefits for our own use immediately.

**Comment by rwgk** [comment]
Sorry for the silence. I'll change this to `_pybind11_conduit_v1_` as soon as I have access to a Linux workstation again. (Unfortunately that may be next week only.)

Thanks for the suggestions!

File: tests/exo_planet_c_api.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2024 The pybind Community.
```
Comment (henryiii) [inline tests/exo_planet_c_api.cpp:1]: I know it shows up a few other places, but I think this is frowned upon by lawyers from discussing this in context of other not-legally defined entities. What does the "The pybind Community" mean?

**Comment by rwgk** [comment]
Thanks a lot @henryiii for jumping in!

I just updated the PR description, to systematically replace `__cpp_conduit__` with `_pybind_conduit_v1_`, and a minor edit related to mypy (since the new method name isn't a dunder name anymore it'll need something more special in mypy).

I'll now play a bit with `py::class_<base>` in one extension, `py::class_<derived, base>` in another, with non-matching `PYBIND11_INTERNALS_ID`, to validate my understanding, before maybe adding an explanation to the PR description. — I don't expect that this will lead to any production code changes.

**Comment by rwgk** [comment]
> I'll now play a bit with `py::class_<base>` in one extension, `py::class_<derived, base>` in another, with non-matching `PYBIND11_INTERNALS_ID`, to validate my understanding, before maybe adding an explanation to the PR description. — I don't expect that this will lead to any production code changes.

Done. @henryiii if this looks good to you, please merge.

I volunteer to backport this to 2.11 and 2.12.

**Review by henryiii — APPROVED** [review]

**Comment by rwgk** [comment]
Continuing the fact finding here, picking up from:

* @rwgk: https://github.com/pytorch/pytorch/pull/136280#issuecomment-2373149065

* @henryiii: https://github.com/pytorch/pytorch/pull/136280#issuecomment-2374340947

> https://docs.python.org/3/reference/lexical_analysis.html#identifiers
>
> Any use of __*__ names, in any context, that does not follow explicitly documented use, is subject to breakage without warning.
>
> Emphasis is from the original docs.

I interpret that as: Use with caution. — That's in line with what Thomas Wouters told me (re-quoting): "They're not reserved, but you do run the risk of clashing with the runtime. There's other third-party packages that use their own dunders."

Concluding from the reference documentation that dunders are "reserved" is a big leap. I believe stating that in isolation is very easily misleading.

I'm not trying to convince anyone to adopt a specific interpretation, but I believe it's important to not jump to conclusions. @henryiii I hope you will edit your comments here and under the pytorch PR to clarify, to make it less likely that people take your interpretation as source of truth.

Beyond that, I'm ok waiting to see if there is any other fallout, although I'm still very much in favor of adopting the dunder convention as soon as possible. With the `pybind11` part in the name there's really no danger of clashing with the runtime. Currently we're the odd ones out with the leading-and-trailing single underscore (AFAIK).

**Comment by henryiii** [comment]
PyTorch was filtering on `__*` to try to pick out only data members. This is incorrect; it just happens to work if the only methods are Python built-in methods, which all are `__*__` methods. If someone added a private data member (`__*`), this would incorrectly filter it out. If someone added a dunder name that stored a member _value_ (like `__dict__`), this is wrong, too. This needs to be fixed in PyTorch, and doesn't mean we were wrong in choosing this name.

The point I'm making is that `__*__` _can't_ be the standard way to define a protocol method, because Python states _any_ (emphasis in original source) usage in any context is subject to breakage without warning. Yes, it doesn't use the word "reserved", but isn't that the definition of reserved? Python _can break_ any usage that is not defined already by Python at any time? How else would you define reserved?

Other libraries absolutely do use `_*_` names for protocols, like `_repr_html_()`. The Jupyter display protocol is one example. They usually are used for single-library or dedicated protocols.

Other libraries also do make their own `__*__` names too, a popular example being NumPy. These are generally used to make cross-library protocols, like `__array_function__` - I suspect looking like it's part of Python itself makes this look "official". While I think that's technically incorrect, and Python would be within their rights to break `__array_function__` tomorrow, it's really unlikely because the fallout would be high.

Once we have a standard mechanism for this, we could propose using a `__*__` name; I think if we make a SPEC, that might be enough. But while it's pybind11 specific, I think we should stick with non-dunder names, unless at least one more (and hopefully more valid) example of breakage comes up. The PyTorch issue would have just happened to work because we would have been filtered the same way as the poor filtering that was already present.

If we redo this, we need to handle reading the old values too, as we've already released and packages have already been built, so it's not as simple as just swapping the name out.

**Pull request merged**